R/get_source_expressions.R

```diff
@@ -39,24 +41,20 @@
 #'   \item{lines}{The [readLines()] output for this file.}
 #' @export
 #' @md
-get_source_expressions <- function(filename) {
+get_source_expressions <- function(filename, lines = NULL) {
   source_file <- srcfile(filename)
-  terminal_newline <- TRUE
 
   # Ensure English locale for terminal newline and zero-length variable warning messages
   old_lang <- set_lang("en")
   on.exit(reset_lang(old_lang))
 
-  source_file$lines <- withCallingHandlers({
-      readLines(filename)
-    },
-    warning = function(w) {
-      if (grepl("incomplete final line found on", w$message, fixed = TRUE)) {
-        terminal_newline <<- FALSE
-        invokeRestart("muffleWarning")
-      }
-    }
-  )
+  source_file$lines <- if (is.null(lines)) {
+    read_lines(filename)
+  } else {
+    lines
+  }
+
+  terminal_newline <- !identical(attr(source_file$lines, "terminal_newline", exact = TRUE), FALSE)
```

Comment:
!identical(., FALSE) seems overly complicated... isn't that just `isTRUE`?

Reply:
It is possible that the input does not have `terminal_newline` at all if the `lines` does not come with the attribute. Therefore I regard all cases other than `FALSE` as `TRUE`, rather than all cases other than `TRUE` as `FALSE`.

Reply:
How about `!isFALSE` then?

Reply:
I didn't use `isFALSE` because I'm not sure if we use backports or something to ensure compatibility with older R versions before `isFALSE` was introduced in 3.5. Do we have a policy on that?

Reply:
Hmm let's just stick with adding a comment

Reply:
Sure